kivy/input/providers/xinput2_core.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * xinput2_core.c - Connect to x and fetch touch events
+ *
+ * Heavily inspired from https://github.com/esjeon/xinput2-touch
+ * => 2013 Eon S. Jeon <esjeon@live.com>
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <X11/Xlib.h>
+#include <X11/extensions/XInput2.h>
+
+#define TOUCH_DOWN 0
+#define TOUCH_MOVE 1
+#define TOUCH_UP 2
+
+static Display *dpy;
+static int xi_opcode;
+
+// Cython API
+typedef struct TouchStruct{
+    int t_id;
+    int state;
+	float  x;
+	float  y;
+} TouchStruct;
+
+typedef int (*touch_cb_type)(TouchStruct *touch);
+touch_cb_type touch_cb = NULL;
+
+void x11_set_touch_callback(touch_cb_type callback) {
+	touch_cb = callback;
+}
+
+
+int init (int windowID){
+	dpy = XOpenDisplay(NULL);
+
+	int devid = 0;
+	Window win = windowID;
+
+	/* check XInput extension */
+	{
+		int ev;
+		int err;
+
+		if (!XQueryExtension(dpy, "XInputExtension", &xi_opcode, &ev, &err)) {
+			printf("X Input extension not available.\n");
+			return 1;
+		}
+	}
+
+	/* check the version of XInput */
+	{
+		int rc;
+		int major = 2;
+		int minor = 3;
+
+		rc = XIQueryVersion(dpy, &major, &minor);
+		if (rc != Success)
+		{
+			printf("No XI2 support. (%d.%d only)\n", major, minor);
+			exit(1);
```

Comment:
Doesn't seem safe to call `exit(1)`. Wouldn't that exit the running process? 
Can we propagate errors so that they can be logged using Kivy `Logger`?